File: swift/Sources/OpenID4VP/HeldMdoc.swift
import CborCose
import MDoc
import SdJwt

/// A held mdoc (ISO 18013-5) exposed to DCQL as a `QueryableCredential`. mdoc claims are a
/// two-level tree `{ namespace: { elementIdentifier: value } }`, so a DCQL claim path is
/// `[namespace, element]` (both strings) — see the mdoc path handling in Dcql parsing.
public struct HeldMdoc: QueryableCredential {
    public let credentialId: String
    public let issuerSigned: IssuerSigned
    public let format = "mso_mdoc"
    public let vct: String? = nil
    public let docType: String?
    public let claims: JsonValue

    public init(credentialId: String, issuerSigned: IssuerSigned) throws {
        self.credentialId = credentialId
        self.issuerSigned = issuerSigned
        self.docType = try issuerSigned.parseMso().docType
        self.claims = .obj(issuerSigned.elements().map { ns, elements in
            (ns, .obj(elements.map { ($0.0, CborJson.toJson($0.1)) }))
        })
    }
}

File: kotlin/openid4vp/src/main/kotlin/com/hopae/eudi/wallet/vp/HeldMdoc.kt
package com.hopae.eudi.wallet.vp

import com.hopae.eudi.wallet.mdoc.IssuerSigned
import com.hopae.eudi.wallet.sdjwt.JsonValue

/**
 * A held mdoc (ISO 18013-5) exposed to DCQL as a [QueryableCredential]. mdoc claims are a
 * two-level tree `{ namespace: { elementIdentifier: value } }`, so a DCQL claim path is
 * `[namespace, element]` (both strings) — see [DcqlEngine] mdoc path handling.
 */
class HeldMdoc(
    override val credentialId: String,
    val issuerSigned: IssuerSigned,
) : QueryableCredential {

    override val format: String = "mso_mdoc"
    override val vct: String? = null
    override val docType: String = issuerSigned.parseMso().docType

    override val claims: JsonValue.Obj = JsonValue.Obj(
        issuerSigned.elements().map { (namespace, elements) ->
            namespace to JsonValue.Obj(elements.map { (id, value) -> id to CborJson.toJson(value) })
        }
    )
}
